### PR TITLE
Feature/simplify cwl

### PR DESCRIPTION
**Pull request by dglaeser**
@joergfunger, this simplifies the cwl workflow to now only expose the geometry file as input, producing the paper at the end. This makes this simple use case easier to understand as the workflow file now looks cleaner and is smaller. The visualization now looks like this: https://view.commonwl.org/workflows/github.com/BAMresearch/NFDI4IngScientificWorkflowRequirements/blob/feature/simplify-cwl/simple_use_case/cwl/wf_run_use_case.cwl

NOTE: This builds upon on #21, which should be merged first. The actual commit history of this branch starts with 4c6d6880ee90f7c0e1f11e18e12eec1a5a7d3055

### CODE REVIEW

**Comment by dglaeser** [comment]
Note that while before (with all the inputs) the workflow was executable from anywhere on the system, now it assumes that it is executed in the cwl folder. I think we have that requirement for other realizations as well, and for the simple use case it shouldn't matter. But in general (e.g. for the Working-Group-User-Story) I think it could be interesting to have workflows that do not assume any specific location to be executed in,  but expose all dependencies.

**Comment by joergfunger** [comment]
IMO the input should not be the geometry file, but either nothing at all (in the case of a real paper you would not produce the same journal paper with different geometries), or some specific values (e.g. the dimension of the specimen when you have to produce the same report for different tests with different geomtries). An alternative would be different parameters of the pde.

**Comment by dglaeser** [comment]
As I said in the comment in the other PR, I don't find our workflow super meaningful. But so far the focus was on keeping it simple and evaluating the differences between the tools. But I'm happy if we can make this one more representative as well. If not, we should be more careful in defining the scope and our "wishes" at least when we start designing the more complex use case.

**Comment by dglaeser** [comment]
But a more general comment, something we discussed many times already I think.. 

> in the case of a real paper you would not produce the same journal paper with different geometries

Also here I think it depends on the use case. You may also be investigating a specific mechanical response (or so) with respect to the specimen geometry, no? Also, in the research/development phase of a project you may also be switching geometries. 

For reproducing a paper, the geometry then should be fixed to the one used in the paper, of course. But which use case does our simple use case represent?

As said last time, I personally think it would be nice to start organizing a bit the use cases and our wishes/recommendations for them and make clear to which group our simple case (or later the more complex one) belongs. Or maybe it's not the simple_use_case that belongs to a specific group but a concrete implementation of it?

That said, I am fine with removing all inputs, then it would probably be the paper reproduction case with everything fixed, no?

**Comment by pdiercks** [comment]
Disclaimer: I am not sure if I understand what you are really after still ....

So CWL allows to define inputs and outputs of the workflow as a whole (might call it top level definition). But, the (now removed from the top level declaration) inputs regarded as specific to this workflow are still inputs and thus dependencies for the single steps of the workflow. 
So all this does is exposing what the user intended to be varied (the top level inputs); Essentially the DAG is simplified. 

In order to understand in which way the input might be varied you need knowledge about the actual workflow. For example, in our simple use case the `.geo` file could still be an input if we are interested in varying the level of refinement of the mesh, but changing the computational domain to a different one might not work at all.

So, the simplified DAG may give a better overview, but to actually use the workflow one has to understand every step anyway.
Also note that most other workflow tools don't have this option to define the top level dependencies (or at least I am not aware of it).

**Comment by joergfunger** [comment]
Maybe this merge request and also the simple example is not the right place to discuss this issue. But from my point of view, each workflow has a set of inputs (potentially empty) and a set of outputs. Even if this is not specifically marked in other tools than CWL, it is still necessary to think about this when others want to reuse the workflow by changing the global workflow inputs (e.g. if a new geometry is included then our fenics script would have to be modified as well, otherwise the workflow does not give a meaningful result for any general geometry). Thus, changing the mesh density would be a valid change the in the geometry file, whereas changing the domain is probably not working due to e.g. the boundary conditions. Thus, it is in general of utmost importance to only specify those parameter as global inputs that are really supposed to be changed (and already thought of by the developer of the workflow). Others are also inputs, but they are static in the sense that they should not be changed (unless they really understand what they are doing) by a new user.

In all tools, you can also specify workflow hiearchically.  For eample, there is one workflow consisting of a single module that takes as input the mesh files and outputs the xdfm. This workflow can be published and resued by others, e.g. in our simple use case. In the longterm future, if someone wants to use that, they do not have to copy paste all the relevant code, but just use (similar to pypi or aiidalab) this workflow module in their own workflow (referencing e.g. a version number of the mesh2xdfm workflow). In case someone changes this (e.g. by adding some functionality or adapting it to new versions of the msh/xdmf file specifications), they would just update the version number in their workflow.

So in summary, just for consistency reasons, I would either set the global input to zero (thus the geo file should not be changed and there is no global input), or add a specific variable (e.g. mesh density) that could be easily changed and that is still working with the rest of the workflow to still ensure a meaningful result.

**Comment by dglaeser** [comment]
> Maybe this merge request and also the simple example is not the right place to discuss this issue.

I think it is ok, let's just make sure that we transfer the relevant parts/outcomes to the documentation or some other, more visible place.

What you describe is exactly what I mean. Collecting our thoughts (in some form of documentation) on what we think are meaningful reusable steps, what are meaningful inputs, etc. It is hard to do on a generic, non-use-case-specific level, but I guess we can come up with some high-level ideas.

For instance, as said last time, one extremum for the gmsh wrapper in cwl would be to simply expose the entire gmsh interface, while the other is fixing it to one concrete example (geofile, dimension, whatever other parameter...). Somewhere in between I guess is a meaningful solution, and maybe we can find some consensus there for our user stories. I fully agree with your reasoning for this workflow @joergfunger, what I mean is to do the same one level of abstraction higher. I know it's difficult without a concrete example, but I guess some high-level heuristics could be formulated, which we could then follow in our implementations to see if they play out well. If not, we reiterate, etc...

**Comment by dglaeser** [comment]
To answer also to @pdiercks:

> So CWL allows to define inputs and outputs of the workflow as a whole

Technically you can do all kinds of things in CWL, I think I have done a bad job explaining what I mean. Let's take the gmsh step,
for instance. CWL executes things in some temporary folder (or maybe even a container if specified), so any file you want to
use from the local file system has to be declared. What I did:

- make geofile an input to the gmsh wrapper
- make geofile an input to the workflow (but with ../source/unit_square.geo as default) and pass that to the wrapper

So the workflow is a hybrid between generic (you can set a geo file), but it defaults to something very
specific to this workflow and the place of execution (it assumes ../source/... to be present)

What I could also to:

- put the default of "../source/unit_square.geo" in the wrapper, making that one wf-specific
- let the workflow have no inputs

or:

- leave geofile as input to the gmsh wrapper
- let the workflow have no inputs and use "../source/unit_square.geo" as input in the `step` section of the mesh generation

In general, both on tool and workflow level you have complete freedom which things you expose as inputs...

**Comment by joergfunger** [comment]
Let's merge this branch and discuss this in the two complex examples.

**Comment by dglaeser** [comment]
> Let's merge this branch and discuss this in the two complex examples.

ok from my side, I just rebased on main to get a nicer history graph.

**Comment by dglaeser** [comment]
@pdiercks, I'm just gonna merge this, I hope it's ok.
